### PR TITLE
[WIP] Handle ObsoleteAttribute correctly with IsByRefLike Struct types

### PR DESCRIPTION
**Pull request by TIHan**
Resolves this issue, https://github.com/Microsoft/visualfsharp/issues/5858.

This PR changes how we handle ObsoleteAttributes for IsByRefLike Struct types, or 'ref structs' in C#. 

Before, we only cared about having any ObsoleteAttribute on a IsByRefLike Struct. But, the correct rule for this is checking for this specific text inside of the ObsoleteAttribute: 
**"Types with embedded references are not supported in this version of your compiler."** 
We call this text, the **ByRefLikeMarker.** This PR now follows this rule.

The PR also emits an ObsoleteAttribute with the **ByRefLikeMarker** above when we create a IsByRefLike Struct; we didn't before and that violates the rule.

All as specified here: https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/span-safety.md#metadata-representation-or-ref-like-structs
> Obsolete attribute with a **known string** will be added to all ref-like structs. Compilers that know how to use ref-like types will ignore this particular form of Obsolete.

The PR is almost complete, these tasks are needed.

- [x] Don't emit ObsoleteAttribute with **ByRefLikeMarker** if the user decides to place an ObsoleteAttribute on the type themselves.
    > In particular, if user wants to actually put an Obsolete or Deprecated attribute on a ref-like type, we will have no choice other than not emitting the predefined one since Obsolete attribute cannot be applied more than once..
- [ ] Tests

// cc @agocke

### CODE REVIEW

**Comment by TIHan** [comment]
I need to cleanup some code. I renamed `CustomAttrs` to `CustomAttributes` and I probably should leave it alone. I will also revisit the implementation a bit more.

File: src/absil/il.fsi [hunk 1]
```diff
@@ -1271,7 +1273,6 @@ and [<NoComparison; NoEquality>]
     member MethodImpls: ILMethodImplDefs
     member Events: ILEventDefs
     member Properties: ILPropertyDefs
-    member CustomAttrs: ILAttributes
```
Comment (dsyme) [inline src/absil/il.fsi]: If you're going to change the name from `CustomAttrs` to `CustomAttributes` then do it everywhere for every type in `il.fsi`/`il.fs` for consistency.  But suggest you do that in a separate PR.

File: src/absil/il.fs [hunk 2]
```diff
@@ -3899,8 +3900,41 @@ let decodeILAttribData (ilg: ILGlobals) (ca: ILAttribute) =
       let v, sigptr = parseVal ty sigptr
       parseNamed ((nm, ty, isProp, v) :: acc) (n-1) sigptr
     let named = parseNamed [] (int nnamed) sigptr
-    fixedArgs, named     
-
+    fixedArgs, named
+    
+[<Literal>]
```
Comment (dsyme) [inline src/absil/il.fs]: Add a comment to this explaining things.  Also no need to make it a literal nor reveal it in the signature, it's better to keep it as clean F# code without use of corner-case interop features like `[<Literal>]`

File: src/absil/il.fs [hunk 3]
```diff
@@ -3899,8 +3900,41 @@ let decodeILAttribData (ilg: ILGlobals) (ca: ILAttribute) =
...
+let ByRefLikeMarker = "Types with embedded references are not supported in this version of your compiler."
+
+type ILTypeDef with
```
Comment (dsyme) [inline src/absil/il.fs]: We don't really use type augmentations like this in `il.fs`.  To keep with the style of the rest of `il.fs` I suggest just using a module function like this:

    let GetFilteredCustomAttributes (ilg: ILGlobals) (td: ILTypeDef) = ....

rather than an instance member.

File: src/absil/il.fs [hunk 4]
```diff
@@ -2016,6 +2016,17 @@ type ILTypeDef(name: string, attributes: TypeAttributes, layout: ILTypeDefLayout
     new (name, attributes, layout, implements, genericParams, extends, methods, nestedTypes, fields, methodImpls, events, properties, securityDecls, customAttrs) =  
        ILTypeDef (name, attributes, layout, implements, genericParams, extends, methods, nestedTypes, fields, methodImpls, events, properties, storeILSecurityDecls securityDecls, storeILCustomAttrs customAttrs, NoMetadataIdx)
 
+    member val CachedFilteredCustomAttributes = None with get, set
```
Comment (dsyme) [inline src/absil/il.fs]: I'm skeptical we should be adding an extra field to ILTypeDef for such a very corner case feature.  @AviAvni worked  very hard to reduce the size of ILTypeDef in 2018....  Many, many of these get instantiated.

Why not just filter on each check through the attributes?

File: src/absil/il.fs (hunk 3 above)
Comment (dsyme) [reply src/absil/il.fs]: On second thoughts I think that this logic should probably be in `AttributeChecking.fs` - it is much more related to the F# compiler (`fsharp\*`) than the raw IL format (`absil\il.fs`).  Is there a specific reason we need to call `GetFilteredCustomAttributes` any other place except in `AttributeChecking.fs`?  Nowhere else checks for `ObsoleteAttribute`....

File: src/fsharp/TypeChecker.fs [hunk 5]
```diff
@@ -10841,7 +10841,7 @@ and TcAttribute canFail cenv (env: TcEnv) attrTgt (synAttr: SynAttribute)  =
                 let tdef = tcref.ILTyconRawMetadata
                 let tref = cenv.g.attrib_AttributeUsageAttribute.TypeRef
                 
-                match TryDecodeILAttribute cenv.g tref tdef.CustomAttrs with 
+                match TryDecodeILAttribute cenv.g tref (tdef.GetFilteredCustomAttributes(cenv.g.ilg)) with
```
Comment (dsyme) [inline src/fsharp/TypeChecker.fs]: I think this call can just be `CustomAttrs` still, no need to filter when looking for `AttributeUsageAttribute`

File: src/fsharp/TastOps.fs [hunk 6]
```diff
@@ -2927,7 +2927,7 @@ let TryBindTyconRefAttribute g (m:range) (AttribInfo (atref, _) as args) (tcref:
         | None -> None
 #endif
     | ILTypeMetadata (TILObjectReprData(_, _, tdef)) -> 
-        match TryDecodeILAttribute g atref tdef.CustomAttrs with 
+        match TryDecodeILAttribute g atref (tdef.GetFilteredCustomAttributes(g.ilg)) with
```
Comment (dsyme) [inline src/fsharp/TastOps.fs]: I think this call can just be `tdef.CustomAttrs`, no need to filter here  as we only filter when searching for `ObsoleteAttribute`

File: src/fsharp/AttributeChecking.fs [hunk 7]
```diff
@@ -248,21 +248,21 @@ let MethInfoHasAttribute g m attribSpec minfo  =
 
 
 /// Check IL attributes for 'ObsoleteAttribute', returning errors and warnings as data
-let private CheckILAttributes (g: TcGlobals) isByrefLikeTyconRef cattrs m = 
+let private CheckILAttributes (g: TcGlobals) cattrs m =
```
Comment (dsyme) [inline src/fsharp/AttributeChecking.fs]: Feels like we can apply the filtering here?

File: src/absil/il.fsi (hunk 1 above)
Comment (TIHan) [reply src/absil/il.fsi]: I'm going to leave it as `CustomAttrs`. Not sure why I wanted to change it to `CustomAttributes`. You are right I would need to change it everywhere if I were to rename it. We can do that later if we decide to do it, but not for this PR as I've decided.

File: src/absil/il.fs (hunk 2 above)
Comment (TIHan) [reply src/absil/il.fs]: I'll see if I can do this without having to expose the literal.

File: src/absil/il.fs (hunk 4 above)
Comment (TIHan) [reply src/absil/il.fs]: I think I agree with you that we shouldn't add something like this to `ILTypeDef`. I think it makes the API confusing.

File: src/fsharp/TypeChecker.fs (hunk 5 above)
Comment (TIHan) [reply src/fsharp/TypeChecker.fs]: Yes, it should just be `CustomAttrs`.

File: src/fsharp/AttributeChecking.fs (hunk 7 above)
Comment (TIHan) [reply src/fsharp/AttributeChecking.fs]: We can. And honestly might be the easiest path to just get what we want to work.

File: src/absil/il.fs (hunk 4 above)
Comment (TIHan) [reply src/absil/il.fs]: Ah I see. It's not just the API. We are also storing more data here.

File: src/fsharp/AttributeChecking.fs (hunk 7 above)
Comment (TIHan) [reply src/fsharp/AttributeChecking.fs]: Ok, I think I remember now. We can't just have the filtering be applied here, we also need it for NameResolution.

**Comment by TIHan** [comment]
I'm going to look at re-doing the implementation here.

My original goal was to filter out the special `ObsoleteAttribute` that we get from an `ILTypeDef`, so it could never be retrieved anywhere, except for when we do IL writing as it needs everything. But, I don't know how feasible that is, or even if it's worth it?

File: src/absil/il.fs (hunk 3 above)
Comment (TIHan) [reply src/absil/il.fs]: We do, we need to filter it in NameResolution as well. That is what originally sparked this PR.

I agree that this is more related to the F# compiler rather than the raw IL format. However, we use the raw IL format types in the compiler directly when making decisions; therefore, they are an integral part of decision making. In a perfect world, the compiler wouldn't know about these raw IL format types and we would provide some sort of abstraction that internally would access these types; that abstraction would be the one responsible for the filtering. Since that is not the case, it's tricky who should be responsible for filtering; I still lean towards `ILTypeDef` being responsible based on this.

**Comment by TIHan** [comment]
After looking at this all day, I couldn't come up with something better. This is tricky to get right without having to modify how we store attributes and/or modify `ILTypeDef` a bit.

However, we can still move forward. I can add the fix for NameResolution quite easily and I already have the code to output the byreflike marker. Perhaps that is the path to least resistance.

**Comment by TIHan** [comment]
With the knowledge I have now over the course of a few months, I severely over-complicated and didn't think through this. @dsyme  is right that we do not want to add any extra field on `ILTypeDef`. We must be vigilant on the cost of adding additional fields to fundamental types used in the compiler. Any extra field yields more memory usage in a significant way.

Will look at this later and either do two things: 
1. Make the fix in NameResolution if the absolete attribute has the `ByRefLikeMarker`
or
2. Filter out the obsolete attribute with the `ByRefLikeMarker` in `ilread.fs` when reading attributes directly.

With that in mind, I'm closing this PR as it must be redone.
